Sherlock/replicator.h

```diff
@@ -0,0 +1,285 @@
+    bool valid_;
+    std::function<void()> done_callback_;
+    std::atomic_bool external_stop_;
+    std::atomic_bool internal_stop_;
+    ScopeOwnedBySomeoneElse<RemoteStream> remote_stream_;
+    F& subscriber_;
+    uint64_t index_;
+    std::thread thread_;
+    std::mutex mutex_;
+    std::condition_variable event_;
+    std::string subscription_id_;
+
+    RemoteSubscriberThread() = delete;
+    RemoteSubscriberThread(const RemoteSubscriberThread&) = delete;
+    RemoteSubscriberThread(RemoteSubscriberThread&&) = delete;
+    void operator=(const RemoteSubscriberThread&) = delete;
+    void operator=(RemoteSubscriberThread&&) = delete;
+
+   public:
+    RemoteSubscriberThread(ScopeOwned<RemoteStream>& remote_stream,
+                           F& subscriber,
+                           uint64_t start_idx,
+                           std::function<void()> done_callback)
+        : valid_(false),
+          done_callback_(done_callback),
+          external_stop_(false),
+          internal_stop_(false),
+          remote_stream_(remote_stream,
+                         [this]() {
+                           TerminateSubscription();
+                           thread_.join();
+                         }),
+          subscriber_(subscriber),
+          index_(start_idx),
+          thread_([this]() { Thread(); }) {
+      static_assert(current::ss::IsEntrySubscriber<F, TYPE_SUBSCRIBED_TO>::value, "");
+      valid_ = true;
+    }
+
+    ~RemoteSubscriberThread() {
+      if (valid_) {
+        if (!subscriber_thread_done_) {
+          TerminateSubscription();
+        }
+        thread_.join();
+      } else {
+        if (done_callback_) {
+          done_callback_();
+        }
+      }
+    }
+
+   private:
+    void Thread() {
+      ThreadImpl();
+      subscriber_thread_done_ = true;
+      if (done_callback_) {
+        done_callback_();
+      }
+    }
+
+    void ThreadImpl() {
+      const std::string url = remote_stream_.ObjectAccessorDespitePossiblyDestructing().url_;
+      bool terminate_sent = false;
+      while (!internal_stop_) {
+        if (!terminate_sent && external_stop_) {
+          terminate_sent = true;
+          if (subscriber_.Terminate() != ss::TerminationResponse::Wait) {
+            return;
+          }
+        }
+        try {
+          HTTP(ChunkedGET(
+              url + "?i=" + current::ToString(index_),
+              [this](const std::string& header, const std::string& value) { OnHeader(header, value); },
+              [this](const std::string& chunk_body) { OnChunk(chunk_body); },
+              [this]() {}));
+        } catch (current::Exception&) {
+        }
+      }
+    }
+
+    void OnHeader(const std::string& header, const std::string& value) {
+      if (header == "X-Current-Stream-Subscription-Id") {
+        std::unique_lock<std::mutex> lock(mutex_);
+        subscription_id_ = value;
+        event_.notify_one();
+      }
+    }
+
+    void OnChunk(const std::string& chunk) {
```

Comment:
This implementation relies on the fact that each chunk is exactly one log entry, which in general is not true. I'd suggest to at least add a comment about this.